File: python/1_basic_blockchain/blockchain.py
"""Copyright [2018] [Charfaoui Younes]

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License."""

# Module 1 - Create a Blockchain

#importing the libraries.
import datetime
import hashlib
import json
from flask import Flask , jsonify

# Building The Blockchain Architecture.
class Blockchain:
    
    def __init__(self):
        self.chain = []
        self.create_block(proof = 1, previous_hash = '0')
        
    # this is a type general block , we can add anything
    def create_block(self, proof, previous_hash):
        block = { 'index' : len(self.chain) + 1,
                 'timestamp' : str(datetime.datetime.now()),
                 'proof' : proof,
                 'previous_hash' : previous_hash }
        self.chain.append(block)
        return block
    
    def get_previous_block(self):
        return self.chain[-1]
    
    # this the main thing about the blockchain, it's define the nonce to use in actual block that satisfay the current target 
    def proof_of_work(self, previous_proof):
        new_proof = 1
        check_proof = False
        while check_proof is False:
            # operation must not be symetrical
            hash_operation = hashlib.sha256(str(new_proof**2 - previous_proof**2).encode()).hexdigest()
            if hash_operation[:4] == '0000':
                check_proof = True
            else:
                new_proof+=1
        return new_proof
    
    # helper function to hash a block
    def hash(self, block):
        encoded_block = json.dumps(block, sort_keys = True).encode()
        return hashlib.sha256(encoded_block).hexdigest()
    
    # helper function to check if the blockchain is valid
    def is_chain_valid(self, chain):
        previous_block = chain[0]
        block_index = 1
        while block_index < len(chain):
            block = chain[block_index]
            if block['previous_hash'] != self.hash(previous_block):
                return False
            previous_proof = previous_block['proof']
            proof = block['proof']
            hash_operation  = hashlib.sha256(str(proof**2 - previous_proof**2).encode()).hexdigest()
            if hash_operation[:4] != '0000':
                return False
            previous_block = block
            block_index += 1
        return True

# The Implementation in the web app.

# Creating web app.
app = Flask(__name__)
   
# Create a blockchain.
blockchain = Blockchain()

# Mining a block.
@app.route('/mine_block', methods = ['GET'])
def mine_block():
    previous_block  = blockchain.get_previous_block()
    previous_proof = previous_block['proof']
    proof = blockchain.proof_of_work(previous_proof)
    previous_hash = blockchain.hash(previous_block)
    block = blockchain.create_block(proof, previous_hash)
    response = {'message': 'Congratulations, you just mined a block!',
                'index': block['index'],
                'timestamp': block['timestamp'],
                'proof': block['proof'],
                'previous_hash': block['previous_hash']}
    return jsonify(response) , 200

# Getting Full blockchain.
@app.route('/get_chain',methods = ['GET'])
def get_chain():
    response = {'chain' : blockchain.chain,
                'length': len(blockchain.chain)}
    return jsonify(response), 200

# Checking if the chain is valid via the helper function.
@app.route('/is_valid', methods = ['GET'])
def is_valid():
    is_valid = blockchain.is_chain_valid(blockchain.chain)
    if is_valid:
        response = {'message': 'All good. The Blockchain is valid.'}
    else:
        response = {'message': 'Houston, we have a problem. The Blockchain is not valid.'}
    return jsonify(response), 200

#Running the app
app.run(host = '0.0.0.0' , port = 5000)

File: kotlin/src/blockchain/BlockChain.kt
/*Copyright [2018] [Charfaoui Younes]

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.*/

package blockchain

import org.json.JSONArray
import org.json.JSONObject

class BlockChain {

    val chain = mutableListOf<Block>()

    constructor()  {
        chain.add(Block(1, 128, "First blockchain.Block", "0000"))
    }

    constructor(blockchain: String){
        val jsonObject = JSONObject(blockchain)
        val blocksArray = jsonObject.getJSONArray("blocks")
        for(i in 0..blocksArray.length()){
            val number = blocksArray.getJSONObject(i).getLong("number")
            val data = blocksArray.getJSONObject(i).getString("data")
            val previousHash = blocksArray.getJSONObject(i).getString("previousHash")
            val hash = blocksArray.getJSONObject(i).getString("hash")
            val nonce = blocksArray.getJSONObject(i).getLong("nonce")
            //chain.add(Block(number,nonce,data,previousHash,hash))
        }
    }

    private fun getLastBlock(): Block = chain[chain.size - 1]

    fun mineBlock(data : String) : Block {
        var nonce = 1
        var check = false
        val lastBlock = getLastBlock()
        val jsonObject = JSONObject()
        jsonObject.put("number" , lastBlock.number + 1)
        jsonObject.put("data" , data)
        jsonObject.put("previousHash" , lastBlock.hash())
        while (!check ){
            val hash = HashUtils.hash(jsonObject.put("nonce", nonce).toString())
            if (hash.get(0) == '0')
                check = true
            else
                nonce++
        }
        val newBlock = Block(lastBlock.number + 1, nonce.toLong(), data, lastBlock.hash())
        chain.add(newBlock)
        return newBlock
    }

    override fun toString(): String {
        val blockChainObject = JSONObject()
        val blocksArray = JSONArray()
        for (block in chain) {
            blocksArray.put(JSONObject(block.toString()))
        }
        blockChainObject.put("blocks" , blocksArray)
        blockChainObject.put("length" , chain.size)
        blockChainObject.put("isValid" , this.isValid())
        return blockChainObject.toString()
    }

    fun isValid() : Boolean {
        var previous = chain[0]
        var currentIndex = 1
        while (currentIndex < chain.size){
            val block = chain[currentIndex]
            if (!block.previousHash.equals(previous.hash)) return false
            if (block.hash[0] != '0') return false
            previous = block
            currentIndex++
        }
        return true
    }
}